### ISSUE TITLE
[Q] Why to 'load a document into the global scope before requiring React'

### ISSUE THREAD

**Issue by kenju** [airbnb/enzyme#933]
[Enzyme's Doc about "JSDOM"](http://airbnb.io/enzyme/docs/guides/jsdom.html) says:

> For the best experience with enzyme, it is recommended that you load a document into the global scope before requiring React for the first time. It is very important that the below script gets run before React's code is run.

However, why it is *recommended* to do so? Why is it *very important* to follow this tips?

We are trying to setup the JSDOM environment in every `beforeEach()/afterEach()` clause (using mocha) instead of `--require setup.js`, because we think that every time the test runs, the global environment should be refreshed to avoid side effects.

However, we realized that there is a difference between below two cases:

1. `require React.js -> require PseudoWindow`
2. `require PseudoWindow -> require React.js`

and wondering why.

Though this is not a bug of Enzyme, and I know that ultimately this relates to how the React.js core works, pardon me for asking this question here. I believe that there are some users facing the same issue.

# Reproducible Example

Here is a reproducible repository with full, working sample codes: https://github.com/kenju/enzyme-jsdom-sample

### Setup JSDOM

Here is a `pseudo_window.js` to setup JSDOM environment:

```js
const { JSDOM } = require('jsdom');

const jsdom = new JSDOM(`<!doctype html><body></body></html>`);
const { window } = jsdom;

function copyPropertyRefs(src, target) {
  Object.keys(src).forEach((property) => {
    if (typeof target[property] === 'undefined') {
      target[property] = src[property];
    }
  });
}

export function setup() {
  global.window = window;
  global.document = window.document;
  global.navigator = {
    userAgent: 'node.js'
  };
  copyPropertyRefs(window, global);
}
```

and we have `setup.js` to simply call the `pseudo_window#setup()` method:

```js
import { setup } from './pseudo_window';
setup();
```

### What to test

Simply testing whether the component can be rendered.

```js
import React from 'react';
import {
  shallow,
  mount,
} from 'enzyme';
import assert from 'power-assert';

import Editor  from '../src/Editor';

describe('Editor', function () {
  it('renderes something with shallow', function () {
    const wrapper = shallow(<Editor />);
    assert.equal(wrapper.length, 1);
  });

  it('renderes something with mound', function () {
    const wrapper = mount(<Editor />);
    assert.equal(wrapper.length, 1);
  });
});
```

## Pattern A - load setup.js before requiring React.js

If you run the below code to run test using `--require setup.js` like [the Enzyme's doc](http://airbnb.io/enzyme/docs/guides/jsdom.html) suggests, you 

```bash
$ $(npm bin)/mocha --require test/setup.js spec/path/to/file.jsx

  Editor
    ✓ renderes something with shallow
    ✓ renderes something with mound (42ms)


  2 passing (431ms)
```

## Pattern B - require React.js and then load setup.js

Here is another pattern: `setup()` inside `beforeEach()`:

```js
import React from 'react';
...
import { setup } from './pseudo_window';
...
describe('Editor', function () {
  // here we call setup()
  beforeEach(function() {
    setup();
  });
  ...
});
```

If we run the test, we see the Warning says: `Warning: This browser doesn't support the `onScroll` event` like this:

```bash
$ $(npm bin)/mocha spec/path/to/file.jsx # without --require test/setup.js


  Editor
    ✓ renderes something with shallow
Warning: This browser doesn't support the `onScroll` event
Warning: This browser doesn't support the `onScroll` event
    ✓ renderes something with mound (54ms)


  2 passing (447ms)
```

## Pattern A vs Pattern B

This warning is raised by `react-dom/lib/ReactDOMComponent.js#enqueuePutListener`:

```js
function enqueuePutListener(inst, registrationName, listener, transaction) {
  ...
  if (process.env.NODE_ENV !== 'production') {
    // IE8 has no API for event capturing and the `onScroll` event doesn't
    // bubble.
    process.env.NODE_ENV !== 'production' ? warning(registrationName !== 'onScroll' || isEventSupported('scroll', true), 'This browser doesn\'t support the `onScroll` event') : void 0;
  }
  ...
}
```

I injected `console.log()` and check the both Pattern A/B cases, and found out that in **Pattern B, `onScroll` event is not injected explicitly**. But, we call `setup()` in `beforeEach()` clause, so how can this happen?

---

Thanks for taking time to read this issue.

**Comment by LukasPirkl** [comment]
This is because React is using following module to check whether it can use DOM
https://github.com/facebook/fbjs/blob/e66ba20ad5be433eb54423f2b097d829324d9de6/packages/fbjs/src/core/ExecutionEnvironment.js
As you can see the check is executed immediately when module is loaded/required and result is saved. So the problem is that the JSDOM must be setup before anything else is required and code from this module is executed (and result saved).

In other words setup JSDOM before test is too late because environment was already checked and DOM was not found.

**Comment by ljharb** [comment]
This is a common pattern in general in JS; all shims and polyfills always need to run before any other code, in all cases.

**Comment by kenju** [comment]
Now I understand, this is just the basic behavior of shims. Thank you to both of you!
